### PR TITLE
fix label bug which harms voc accuracy

### PR DESCRIPTION
**Pull request by shenh10**
the problem is stated at #6809 and #6848

### CODE REVIEW

**Comment by shenh10** [comment]
This version is directly forked from the latest master branch, the only change in my version is just one-line of code. I have successfully runed with my Ubuntu 14.04.  I am quite sure it's the bug of master branch but nothing related to my fix. Could you double check on this problem?

**Comment by shenh10** [comment]
@mxnet-jenkins

File: example/rcnn/rcnn/io/rcnn.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def sample_rois(rois, fg_rois_per_image, rois_per_image, num_classes,
     # select labels
     labels = labels[keep_indexes]
     # set labels of bg_rois to be 0
-    labels[fg_rois_per_this_image:] = 0
+    labels[overlaps[keep_indexes] < config.TRAIN.FG_THRESH] = 0
```
Comment (piiswrong) [inline example/rcnn/rcnn/io/rcnn.py]: @precedenceguo

**Comment by shenh10** [comment]
As stated in #6848, to remain the significance of FG_FRACTION parameter, I have changed the padding pool from all RoIs to negative RoIs only. To validate the change, I have retrained a model with dataset voc2007+voc2012 on ResNet. Testing mAP on voc2007 is 0.7929. @Zehaos

**Comment by Zehaos** [comment]
@shenh10 Which backbone do you use? Resnet101 or Resnet50?

**Comment by shenh10** [comment]
@Zehaos It's ResNet-101

**Comment by precedenceguo** [comment]
Great work!
The padding of background rois is to make sure that selected rois reach the fixed batch size. It is possible that there is not enough foreground rois selected (or background rois), where the input rois would be less than the fixed batch size.
However, as you discovered, the padding can come from any roi, regardless of background or not, which conflicts the implicit assumption that any roi after fg_rois_per_image is background (hence the hard set label to zero). This constitutes label conflicts that harms the accuracy.
After the proposed change, the implicit assumption that background rois follows foreground rois is preserved.

I think we need to rerun the jenkins ci for changes only in examples. @piiswrong

**Comment by shenh10** [comment]
@precedenceguo Thank you, very clear explanation. I don't know what's the testing issue in the integration testing phase, hopefully you can do integration test on examples or rcnn module individually.

**Pull request merged**